src/metabase/search/filter.clj

```diff
@@ -135,6 +131,38 @@
     (sql.helpers/where query [:= (search.config/column-with-model-alias model :authority_level) "official"])
     (sql.helpers/where query false-clause)))
 
+;; Created at filters
+
+(defn- default-created-at-filter-clause
+  [model created-at]
+  (let [{:keys [start end unit]} (try
+                                  (params.dates/date-string->range created-at)
+                                  (catch Exception _e
+                                    (throw (ex-info (tru "Failed to parse created at param: {0}" created-at) {:status-code 400}))))
+        start                    (when start (-> start u.date/parse (u.date/range unit) :start))
+        end                      (when end   (-> end u.date/parse (u.date/range unit) :end))
```

Comment:
this is basically how we're processing the date param.  You could find we're doing the same thing in the `optimize-temporal-filter` middleware.

```clojure
(metabase.query-processor.middleware.optimize-temporal-filters/optimize-filter  [:between [:field 1 nil]
                                                                                 [:absolute-datetime #t "2000-01-01" :day]
                                                                                 [:absolute-datetime #t "2000-01-10" :day]])
[:and
 [:>= [:field 1 nil] [:absolute-datetime #t "2000-01-01" :default]]
 [:< [:field 1 nil] [:absolute-datetime #t "2000-01-11" :default]]]
```

The end shifting happens at this middleware.


Why? I don't have a strong reason, my guess is we want to keep `date-string->range` simple and don't include the behavior logic there.

It makes more sense for
```
(date-string->range "2000-01-01~2000-01-10")
```

to returns 
```
{:start "2000-01-01", :end "2000-01-10"}
```

than 
```
{:start "2000-01-01", :end "2000-01-11"}
```

I think.

if we want to change this, we'll touch the QP and basically how we process datetime param.

Reply:
Okay, it makes sense to leave the QP alone. But fixing the issue in this way is complicating matters further and likely to produce bugs. I already found one in your tests:

https://github.com/metabase/metabase/blob/6e4f343b25ed590b216c514f67523b53c36b57b4/test/metabase/search/filter_test.clj#L99-L100

The expected result should actually be `[[:>= :card.created_at #t "2016-04-18T10:00"] [:< :card.created_at #t "2016-04-24T11:00"]]`

Ideally this logic that adjusts the output of date-range should be reusable and separated from global search filters. I think using the `{:inclusive-end? false}` option with `params.dates/date-string->range` was actually meant for this use case, and I tried using it instead of using `(u.date/range <> unit)`. It mostly works but it broke some of your tests in `build-created-at-filter-test`.
It's definitely a little broken, but maybe it can be fixed. Here's some examples I tried:
```clojure
  (params.dates/date-string->range "past1days" {:inclusive-end? false})
  ;; => {:start "2023-08-10", :end "2023-08-11", :unit :day}
  ;; ✅

  (params.dates/date-string->range "past1days~" {:inclusive-end? false})
  ;; => {:start "2023-08-10", :end "2023-08-12", :unit :day}
  ;; ✅

  (params.dates/date-string->range "past1months" {:inclusive-end? false})
  ;;  => {:start "2020-07-01", :end "2020-08-31", :unit :month}
  ;; ❌
  ;; should be
  ;;  => {:start "2020-07-01", :end "2020-08-01", :unit :month}
```

The `{:inclusive-end? false}` option is only used once in the code, by the mongo driver: https://github.com/metabase/metabase/blob/6e4f343b25ed590b216c514f67523b53c36b57b4/modules/drivers/mongo/src/metabase/driver/mongo/parameters.clj#L77

Given it is used only here, I wouldn't be surprised if the bug I found above is happening to mongo field filters and needs fixing.

What do you think about fixing the bugs with this option, and using it instead?

Reply:
Here's the `git diff` output for using the `{:includsive-end? false}` option with your code. It includes a bug fix where `:day` was used regardless of the unit, now the unit is used

```diff
diff --git a/src/metabase/driver/common/parameters/dates.clj b/src/metabase/driver/common/parameters/dates.clj
index 02e844fbfb..e236d7eb2b 100644
--- a/src/metabase/driver/common/parameters/dates.clj
+++ b/src/metabase/driver/common/parameters/dates.clj
@@ -367,10 +367,10 @@
       (m/update-existing :start #(if inclusive-start?
                                    %
                                    ;; seems wrong to add unit day by default, what is the filter unit is hour?
-                                   (u.date/add % :day -1)))
+                                   (u.date/add % (:unit temporal-range) -1)))
       (m/update-existing :end #(if inclusive-end?
                                  %
-                                 (u.date/add % :day 1)))))
+                                 (u.date/add % (:unit temporal-range) 1)))))
 
 (def ^:private DateStringRange
   "Schema for a valid date range returned by `date-string->range`."
diff --git a/src/metabase/search/filter.clj b/src/metabase/search/filter.clj
index 93e56eaadd..22ddc0c708 100644
--- a/src/metabase/search/filter.clj
+++ b/src/metabase/search/filter.clj
@@ -134,12 +134,12 @@
 
 (defn- default-created-at-filter-clause
   [model created-at]
-  (let [{:keys [start end unit]} (try
-                                  (params.dates/date-string->range created-at)
+  (let [{:keys [start end]} (try
+                                  (params.dates/date-string->range created-at {:inclusive-end? false})
                                   (catch Exception _e
                                     (throw (ex-info (tru "Failed to parse created at param: {0}" created-at) {:status-code 400}))))
-        start                    (when start (-> start u.date/parse (u.date/range unit) :start))
-        end                      (when end   (-> end u.date/parse (u.date/range unit) :end))
+        start                    (some-> start u.date/parse)
+        end                      (some-> end u.date/parse)
         created-at-col           (search.config/column-with-model-alias model :created_at)
```

Reply:
nice! this is definitely better. I initially thought that having inclusive-end? = false` will shift the date by `-1` unit, not `+1` so that's why I didn't use it at first.

anyway I've updated it so that we don't have to handle the date at `search.filter` ns, `date-string->range` will return the desired range now.